File: Cpp/LeetCode152.MaximumProductSubarray.cpp
// Time Complexity: O(N)
// Space Complexity: O(1)

#include <vector>

class Solution {
public:
    int maxProduct(std::vector<int>& nums) {
        int N = nums.size();
        
        int prevMax = nums[0];
        int prevMin = nums[0];
        int ans = prevMax;
        
        for (int i = 1; i<N; i++)
        {
            int tempPrevMax = prevMax; // since prevMax will be updated before prevMin
            prevMax = std::max(tempPrevMax*nums[i], prevMin*nums[i]);
            prevMax = std::max(nums[i], prevMax);
            prevMin = std::min(tempPrevMax*nums[i], prevMin*nums[i]);
            prevMin = std::min(nums[i], prevMin);
            ans = std::max(prevMax, ans);
        }
        return ans;
    }
};

File: Python/LeetCode152.MaximumProductSubarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        N = len(nums)

        if N == 1: return nums[0]

        maxP = nums[0]
        minP = nums[0]
        ans = nums[0]

        for i in range(1, N):
            temp = maxP
            maxP = max(nums[i], nums[i] * maxP)
            maxP = max(maxP, nums[i] * minP)
            minP = min(nums[i], nums[i] * minP)
            minP = min(minP, nums[i] * temp)
            ans = max(maxP, ans)

        return ans